how_to_eurec4a/cloudmasks.md

```diff
@@ -4,136 +4,261 @@ jupytext:
 
 ```{code-cell} ipython3
 %pylab inline
 ```
 
-## Get data
-* To load the data we first load the EUREC4A meta data catalogue. More information on the catalog can be found [here](https://github.com/eurec4a/eurec4a-intake#eurec4a-intake-catalogue).
-
 ```{code-cell} ipython3
 import eurec4a
+import numpy as np
+import xarray as xr
+import matplotlib as mpl
+mpl.rcParams['font.size'] = 12
+
+from mpl_toolkits.axes_grid1 import make_axes_locatable
+from matplotlib.colors import LogNorm
+```
+
+We generally don't support ignoring warnings, however, we do so in this notebook to suppress warnings originating from zero division. Unfortunately, we coudn't find another way to handle the warnings. If you have an idea, please make a pull request and change it :)
+
+```{code-cell} ipython3
+import warnings
+warnings.filterwarnings("ignore")
+```
+
+## Cloud fraction functions
+We define some utility functions to extract (circle) cloud cover estimates from different datasets assuming that the datasets follow a certain flag meaning convention.  
+
+In particular, we define a **minimum cloud fraction** based on the cloud mask flag `most_likely_cloudy` and a **maximum cloud fraction** combining the flags `most_likely_cloudy` and `probably_cloudy`. The cloud mask datasets slightly vary in their flag definition for cloud free conditions and their handling of unvalid measurements which is taken care of by the following functions.
+
+```{code-cell} ipython3
+def _isvalid(da):
+    meanings = dict(zip(da.flag_meanings.split(" "), da.flag_values))
+    return ~(np.isnan(da) | ("unknown" in meanings and da==meanings["unknown"]))
+
+def _cloudy_max(da):
+    meanings = dict(zip(da.flag_meanings.split(" "), da.flag_values))
+    return (da==meanings["most_likely_cloudy"]) | (da==meanings["probably_cloudy"])
+
+def _cloudy_min(da):
+    meanings = dict(zip(da.flag_meanings.split(" "), da.flag_values))
+    return da==meanings["most_likely_cloudy"]
+
+def cfmin(ds):
+    sumdims = [d for d in ds.cloud_mask.dims if d != "time"]
+    return (_cloudy_min(ds.cloud_mask).sum(dim=sumdims)
+            / _isvalid(ds.cloud_mask).sum(dim=sumdims))
+
+def cfmax(ds):
+    sumdims = [d for d in ds.cloud_mask.dims if d != "time"]
+    return (_cloudy_max(ds.cloud_mask).sum(dim=sumdims)
+            / _isvalid(ds.cloud_mask).sum(dim=sumdims))
+
+def correct_VELOX(ds):
+    return ds.assign(CF_min=ds.CF_min.where((ds.CF_min >=0 ) & (ds.CF_min <= 1)),
+                     CF_max=ds.CF_max.where((ds.CF_max >=0 ) & (ds.CF_max <= 1)))
+
+def ensure_cfminmax(ds):
+    if "CF_min" not in ds:
+        ds = ds.assign(CF_min=cfmin)
+    if "CF_max" not in ds:
+        ds = ds.assign(CF_max=cfmax)
+    return correct_VELOX(ds)
+
+def load_cloudmask_dataset(cat_item):
+    return ensure_cfminmax(xr.concat([v.get().to_dask().chunk()
+                                      for _, v in cat_item.items()], dim="time"))
```

Comment:
This one loads the individual datasets (and in particular coordinate indices) for each flight day in parallel. That should have a bit of an effect on data loading speed.
```suggestion
from multiprocessing.pool import ThreadPool

def load_cloudmask_dataset(cat_item):
    # load in parallel as this function is mainly limited by the network roundtrip time
    p = ThreadPool(20)
    return ensure_cfminmax(xr.concat(list(p.map(lambda v: v.get().to_dask().chunk(), cat_item.values())),
                                     dim="time"))
```